documentation-support/src/main/java/tech/picnic/errorprone/documentation/BugPatternTestExtractor.java

```diff
@@ -0,0 +1,179 @@
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.util.TreeScanner;
+import com.sun.tools.javac.util.Context;
+import java.util.ArrayList;
+import java.util.List;
+import org.jspecify.annotations.Nullable;
+import tech.picnic.errorprone.documentation.BugPatternTestExtractor.BugPatternTestDocumentation;
+
+/**
+ * An {@link Extractor} that describes how to extract data from a test that tests a {@code
+ * BugChecker}.
+ */
+@Immutable
+final class BugPatternTestExtractor implements Extractor<BugPatternTestDocumentation> {
+  private static final Matcher<MethodTree> JUNIT_TEST_METHOD =
+      hasAnnotation("org.junit.jupiter.api.Test");
+
+  // XXX: Improve support for correctly extracting multiple sources from a single
+  // `{BugCheckerRefactoring,Compilation}TestHelper` test.
+  @Override
+  public BugPatternTestDocumentation extract(ClassTree tree, Context context) {
+    VisitorState state = VisitorState.createForUtilityPurposes(context);
+    CollectBugPatternTests scanner = new CollectBugPatternTests(state);
+
+    tree.getMembers().stream()
+        .filter(MethodTree.class::isInstance)
+        .map(MethodTree.class::cast)
+        .filter(m -> JUNIT_TEST_METHOD.matches(m, state))
+        .forEach(m -> scanner.scan(m, null));
+
+    String className = tree.getSimpleName().toString();
+    return new AutoValue_BugPatternTestExtractor_BugPatternTestDocumentation(
+        className.substring(0, className.lastIndexOf("Test")),
+        scanner.getIdentificationTests(),
+        scanner.getReplacementTests());
+  }
+
+  @Override
+  public boolean canExtract(ClassTree tree, VisitorState state) {
+    String className = tree.getSimpleName().toString();
+    if (!className.endsWith("Test")) {
+      return false;
+    }
+
+    ScanBugPatternTest scanBugPatternTest = new ScanBugPatternTest();
+    scanBugPatternTest.scan(tree, state);
+
+    String bugPatternName = className.substring(0, className.lastIndexOf("Test"));
+    return scanBugPatternTest.hasTestUsingClassInstance(bugPatternName);
+  }
+
+  private static final class ScanBugPatternTest extends TreeScanner<@Nullable Void, VisitorState> {
+    private static final Matcher<ExpressionTree> BUG_PATTERN_TEST_METHOD =
+        staticMethod()
+            .onDescendantOfAny(
+                "com.google.errorprone.CompilationTestHelper",
+                "com.google.errorprone.BugCheckerRefactoringTestHelper")
+            .named("newInstance");
+
+    private final List<String> encounteredClasses = new ArrayList<>();
+
+    boolean hasTestUsingClassInstance(String clazz) {
+      return encounteredClasses.contains(clazz);
+    }
+
+    @Override
+    public @Nullable Void visitMethodInvocation(MethodInvocationTree node, VisitorState state) {
+      if (BUG_PATTERN_TEST_METHOD.matches(node, state)) {
+        MemberSelectTree firstArgumentTree = (MemberSelectTree) node.getArguments().get(0);
```

Comment:
While not likely in our code base, this could cause a `ClassCastException`. Can be fixed by augmenting the `Matcher`.